### ISSUE TITLE
output of long lines from grep truncated in mintty terminal in ssh session

### ISSUE THREAD

**Issue by nngo** [mintty/mintty#840]
I am running mintty 2.9.4 (x86_64-pc-msys) from Git Bash for Windows.

In my mintty terminal (with terminal type set to default of xterm), I ssh to a remote machine.
There I run the following command:
```shell
$ ssh remote-host

remote-host $ ps -aux | grep java
 <nothing>

remote-host $ ps  -aux | grep java | wc -l
  5
```
When my terminal is not wide enough, the output of the long lines are truncating somewhere and it outputs missing lines (sometimes none), this is with a terminal width of 80 chars.

When I expand the terminal's width to 100 chars, I get one line out of expected 5 lines.
When I expand the terminal's  width to 110 chars, I get the expected 5 lines.

I can reproduce this problem consistently when ssh to a remote host in my mintty.
On my local machine, mintty does not appear to not truncate the output to the terminal.

**Comment by mintty** [comment]
For such weird issues, a number of reasons are possible, most unlikely the terminal.
Therefore it is necessary to provide a reproducible test case.
Please upload a file that reproduces the effect when simply output with `cat` (locally preferred).
It is not unlikely that you discover the problem while producing a test case...

**Comment by nngo** [comment]
I ran `ps -aux | grep java > test.txt` with my mintty width at 80 and it created a zero length file.
Did the same again again at width of 100 and it had one line.
Then yet again with width of 110 and it had all five lines.

So maybe issue is upstream of mintty, somewhere between standard output (pipe) of `ps` to `grep` in my ssh session?

**Comment by nngo** [comment]
The remote host is running FreeBSD 11.2, and this happens only inside a jail. Outside the jail, the ps -aux | grep produces the right output and wraps. Closing this issue.

**Comment by mintty** [comment]
Since the redirected output of ps and grep into test.txt is not going through mintty and has the effect already, this is obviously not a mintty issue.
I'd first check whether the versions of ps and grep are actually from /bin or maybe you have any scripts intercepting the invocations. Don't know what a jail is. If you find the culprit, please report back, for curiosity.

**Comment by BrianInglis** [comment]
% ps aux would be a normal BSD ps command like $ ps -efl but $ ps -aux gives me ps: user x unknown
https://en.wikipedia.org/wiki/FreeBSD_jail is a super restricted shell with alternate root e.g. ~user/, few permissions, and restricted commands (e.g. no cd outside or above alternate root)
OP does not state what his jail shell  ;^> is set to

**Comment by nngo** [comment]
Using Bash 4.4.19 on the remote host and it's jail.

**Comment by nngo** [comment]
So, the source issue is actually in FreeBSD's `ps` command which was not outputting stuff when the terminal width is less than the output width. A workaround is the use the `-w` twice for unlimited width (e.g. `ps -auxww` or `ps auxww`).

See FreeBSD Bug 217159 - ps default to 79 characters can break ps|grep in shell scripts
https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=217159

**Comment by nngo** [comment]
Right, the [ps | grep is broken on FreeBSD](http://deepix.github.io/2016/10/10/psww.html) blog post from Oct 2016 gives a very good detail of FreeBSD 11.0's issue that he files in Feb 2017, they finally accepted his patch into their baseline in Mar 2018, but has not been pushed out to FreeBSD 11.2.

Leaving out the dash makes no difference. 'ps aux > width80-no-dash.txt' with terminal width at 80, truncates output at 80 chars. Also tried `ps -aux < /dev/null > width80-no-input.txt` and that was truncated too at 80 chars. I used `wc -L  file` to find the max width of the text file. I also inspected it via `vi`, `more`, and `less` to validate the output width.

One further funkyness of this issue on FreeBSD 11.2, with a terminal width of 50 chars, `ps` truncates at 133 chars, at  width of 74 it truncates at 157, with a minimum terminal width of 13, it truncates at 98 chars.
```shell
$ wc -L ps*                                                            
      98 ps13.txt
    133 ps50.txt
    153 ps70.txt
    157 ps74.txt  <-- what happened starting with terminal width of 75 or less?
      80 ps75.txt
      80 ps80-no-dash.txt
      80 ps80-no-input.txt
      80 ps80.txt
  1311 ps80-ww.txt   <-- with -ww for unlimited width
```
